### ISSUE TITLE
Permanent process memory growth when DiskQ is used

### ISSUE THREAD

**Issue by schenksj** [aol/moloch#1127]
If a moloch process needs to use the diskQ any memory used for the queue is permanently reserved and held by the capture process due to `mlockall(MCL_FUTURE)`.  This can be a big problem for high-volume capture processes as it can cause the host machine to unnecessarily lose access to a large amount of memory.  With `MCL_FUTURE` the process will always occupy the maximum amount of memory it has utilized, but diskQ can have very occasional short-lived spikes (during disk maintenance, an odd long-running yara scan that causes a packet queue to back up and clear quickly, etc...).

To mitigate against this, I'd like to propose that an option be added to set a static mlock amount at startup.  Here is some pseudo-code for how that might work:

```
void moloch_mlockall_init()
{
 <snip>
    int result;
    if (boundedMlockSize > 0) {
       void* ptr = MOLOCH_SIZE_ALLOC0(bound, boundedMlockSize);
       result = mlockall(MCL_CURRENT);
       MALLOC_SIZE_FREE(bound, ptr);
    } else {
       result = mlockall(MCL_FUTURE | MCL_CURRENT);
 <snip>
}
```

Very interested in the communities feedback...

**Elasticsearch version**:

**Moloch version**:

**OS name and version**:

**How was Moloch built/installed:** (rpm, deb, easybutton, ...)

**Provide logs, stack traces and steps to reproduce**:

**Comment by awick** [comment]
Pretty sure this isn't needed - https://linux.die.net/man/3/munmap

```Any memory locks (see mlock() and mlockall()) associated with this address range shall be removed, as if by an appropriate call to munlock().```

Moloch does keep 16 buffers per thread available, maybe that is what you are seeing?

**Comment by schenksj** [comment]
Cool, I didn't know about that.  Are you proposing that some monitor calls "munlock(); mlockall()" at some triggered interval?  Seems a little risky?

The process growth I'm seeing takes my processes to 45 gb or more from a steady state of 6-12gb (yes, I wish I had smaller machines but I don't)

**Comment by awick** [comment]
>

> Cool, I didn't know about that. Are you proposing that some monitor calls "munlock(); mlockall()" at some triggered cadence/criteria? Seems a little bit of risk that you might not get your memory back when you try to re-lock?

No, I'm saying that diskq buffers use mmap and munmap so at least as I read the above, munmap takes care of any munlock.  However up to 16 buffers per thread are kept and not munmaped and just reused when not in used.  So you can calculate the max with `16*packetThreads*pcapWriteSize` and unless you've changed the code the max buffers it keeps around (that aren't in use) is ~3G if you've maxed everything.


> The process growth I'm seeing takes my processes to 45 gb or more from a steady state of 6-12gb (yes, I wish I had smaller machines but I don't). Would this be supportive of the 16 diskQ buffers theory? I have 12 packet threads (which I'll probably reduce now that the Yara thread-starvation #1128 issue has from has been resolved-ish)

No, I would guess this is more from the number of sessions being monitored, and maybe from any snf/afpacket buffers.

**Comment by schenksj** [comment]
Interesting, I'll look deeper into this when I can get back to my project.  It would be cool to have some process memory accounting for the big regions of malloc activity...
